### ISSUE TITLE
Delay in Execution for Robot Tests Increased Massively from v3.2.1 to v4.0.1

### ISSUE THREAD

**Issue by rmccarth-bosch** [robotframework/robotframework#3950]
This may not be a defect but more of a general observation -- we have a suite that used to take approximately 15 minutes to execute a --dryrun command line option on with version 3.2.1.
Once we bumped to version 4.0.1, the --dryrun speed now takes up to 1.5 hours. 
I looked at the change log and don't see anything obvious but is there a functionality difference that would account for this? Are there any known ways of getting higher efficiency out of this option in versions above 4.0? Thanks!

#3769 Was the only mention of --dryrun recently..

**Comment by pekkaklarck** [comment]
Even 15 minutes sounds like a _really_ long `--dryrun` time but 1.5 hours is definitely lot worse. Can you confirm that the only difference between execution rounds is RF version?

Unfortunately I don't have any idea what could cause such a massive slowdown and without being able to run your tests I cannot really debug this further either. There are some things that could be done but they require help from you:

- If you could create an example demonstrating the problem that I could run locally it would allow me to debug this. The example needed to as simple as possible and obviously shouldn't contain any confidential information or require anything too special to be executable.
- You could test with different RF versions to see could we find when the problem occurred. You should start from RF 4.0 alpha 1. If this approach doesn't provide enough granularity, you could clone this project and use e.g. [git bisect](https://git-scm.com/docs/git-bisect) to find the exact commit causing the problem.

**Comment by rmccarth-bosch** [comment]
@pekkaklarck I should of specified **suites** -- we have collections of Robot tests broken up by directories. Our 'main' set of suites takes about those specified times. There are about 300 discrete '.robot' tests with over 4500 test cases.  
They range in complexity from direct keyword or functions calls to nested upon nested upon nested etc. keyword calls.. 
Let me start with some basic debugging and see if I can pin it to a specific version or commit like you suggested.
If I still have issues, let me see what kind of approval I can get to better help you. 

I'll be in touch again soon! Thanks.

**Comment by pekkaklarck** [comment]
If you can analyze which suites are particularly slow with the new version compared to the old, perhaps you could create similar example without your real data.

Actual library keywords shouldn't matter because they aren't executed on dry-run. Libraries are imported, though, so that's something that could also be slower nowadays.

**Comment by theodore86** [comment]
Indeed, the --dryrun has an performance degradation, it doubles (x2) the execution time compare to v3.2.2 based on my tests ( i am using the Pabot library for parallel execution). The affected version seems to be the v4.0rc1, at least until v4.0.b3 the execution time was pretty close to v3.2.2.

I will dig more into it to find out what exactly is causing this performance penalty.

Thanks!

**Comment by pekkaklarck** [comment]
It's great if someone can investigate this further. I'm be happy to debug this myself as well if someone could create as simple example demonstrating the performance problem that I can run locally.

**Comment by myjniak** [comment]
@theodore86 once this is resolved, dryrun execution time could also be significantly improved with this idea:
https://github.com/robotframework/robotframework/issues/3241
But I'm afraid, the priority never got high enough for Robot Framework team to do it.

Anyway, I'm looking forward for this issue to be solved, because it blocks our upgrade to Robot 4.X.X.

**Comment by rmccarth-bosch** [comment]
@myjniak That is very interesting! In #3241, it mentioned the use of 'complex keywords' which we have been using for years and only noticed this improvement drop after the switch to V4.0.1.

@pekkaklarck  I would like to note that this 'delay' before execution is also present when running a Robot as is test **without** the --dryrun option. I am not sure if this affects the priority of this issue. 
In general, our suites would instantly jump to Suite Setup step after mere 100's of milliseconds after kicking them off pre 4.0.1.
Now, this 'delay time' is up to ~20 seconds. 

I am going to change the title of this defect to reflect more recent findings from our testing team. If I can help provide data, please let me know and I can arrange a call or something. Thank you!

**Comment by myjniak** [comment]
I've tried this:

```
*** Test Cases ***

A
    FOR    ${i}     IN RANGE     50000
        Log     ${i}
    END
```

And the performance of RF 3.2.1 and 4.1 was comparable.
I think it'd be best to just find the simplest code possible to reproduce the speed drop you're experiencing.

I've also tried running our complex set of suites as dryrun (1338 tests in ~50 suites) (5 executions with each RF version) and the execution time increased from ~41s (min 38.5s max 45s) to ~45s (min 44s max 50s).

So:

- our team is actually not blocked by this.
- our suites seem to be also affected by the slowdown you found, but only a bit

In conclusion, the problem seems to be real, but it probably requires using a very specific Keyword / logic structure, that your tests do and my tests almost don't.

**Comment by rmccarth-bosch** [comment]
@myjniak You are absolutely correct and I get the exact same results when using simple built-in calls. Those work quite efficiently regardless of RF version. Once you import custom libraries or resource files that have nested on nested keywords is once this problem reveals itself in the post RF 4.0.1 version.

**Comment by myjniak** [comment]
I wonder if it's related to error handling, like Pekka presented here: https://github.com/robotframework/robotframework/issues/4053

**Comment by pekkaklarck** [comment]
Has anyone been able to debug this any further? An example demonstrating performance degradation would be especially helpful because it would make it possible for me to debug the problem on my machine. It would be good also if you could test with RF 4.1.2. There are some performance enhancements and especially pretty big reduction to memory usage.

**Comment by simplyviks** [comment]
We are using RF 4.1.2 and hereby confirm the performance degradation present in 4.1.2 also. i'll to try to add an working example.

**Comment by pinelopiskl** [comment]
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/pinelopi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/pinelopi/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/pinelopi/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EL style='tab-interval:36.0pt;word-wrap:break-word'>


<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'>Below you
can see some comparisons made between 3.1.2 and 4.1.2 in a simple loop that
logs “hello world”.<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'>As you can
see there are two differences in the time between the two versions.<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'>e.g for
running 100 loops the time difference is:<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'>1’47”
versus 5’28”: the total execution-time logged in ride<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'>0’21”
versus 2’10”: the execution-time logged in robot logs<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=EN-US style='color:#2F5496;mso-ansi-language:
EN-US'><o:p>&nbsp;</o:p></span></p>



  | 100 loops | 10000 loops
-- | -- | --
Robot Version | Ride log time | Robot log time | Ride log time | Robot log time
3.1.2 | 1’47’’ | 0’21 | 3’22’’ | 1’26’’
4.1.2 | 5’28’’ | 2’10 | 11’18’’ | 7’29’’



<p class=MsoNormal><span lang=EN-US style='mso-ansi-language:EN-US'><o:p>&nbsp;</o:p></span></p>


</body>

</html>

Robot log 3.1.2 for 100 loops

<img width="2240" alt="Robot_log_100_loops_3 1 2" src="https://user-images.githubusercontent.com/84320048/168598088-f70bdca8-1a8a-457f-9a1c-aed38a3800e4.png">

Robot log 4.1.2 for 100 loops

<img width="2240" alt="Robot_log_100_loops_4 1 2" src="https://user-images.githubusercontent.com/84320048/168598177-18cbeeba-f18e-4217-88b5-dc645707f996.png">
Ride log 4.1.2 for 100 loops

<img width="1427" alt="Ride log 100 loops 4 1 2" src="https://user-images.githubusercontent.com/84320048/168598239-1bf6b0fd-246a-4ff9-932d-8bbbba5ecbc8.png">

**Comment by pekkaklarck** [comment]
We'd need measurements with the latest code to see what's the current status. Measurements should also be done without RIDE or any such runner. Notice also that if you need to run 10000 loop iterations, doing that in Python is probably a good idea anyway.

**Comment by rmccarth-bosch** [comment]
@pekkaklarck I was never able to show it with 'test data' but in one of my earlier comments I mentioned that it has to be done with the building of nested resource files which slowed down performance. Not these long looping iterations.

**Comment by pekkaklarck** [comment]
RF 5.1 will contain a quite big performance enhancement related to checking do keywords have special `robot:<name>` tags. The addition of `robot:continue-on-failure` was the reason for the problems reported in #4053 and these special tags may have contributed to the problems reported in this issue as well. It would be great if someone could test that latest changes now or at least when we RF 5.1 preview releases out.

**Comment by pekkaklarck** [comment]
To be able to debug the problems reported as part of this issue, I'd really need a concrete example. If nobody can create one, it's probably best to just close this issue.

**Comment by pekkaklarck** [comment]
Has anyone tested do RF 5.1 preview releases help with this situation? As mentioned above, RF 5.1 will contain some nice performance enhancements and the latest beta already has everything that was planned to be added.

**Comment by pekkaklarck** [comment]
I'm not aware of others reporting performance regressions with RF 6. In general performance in RF 6 should be better than in RF 5 or 4, but there certainly can be special cases where that's not true. For example, you seem to be using `Run Keyword` a lot which isn't that typical

Could you @dinualx submit a new issue about performance issues with RF 6 compared to RF 5? Please also explain a bit more about your usage. It would be best if you could create a simple example we could run locally that demonstrates the problem. Without such example its' actually nearly impossible for us to do anything.

This issue is about performance regression between RF 3.2 and 4.0. There haven't been new comments related to that since June so I guess the issue can be closed.
